readme.md
# Prysmo

A service gateway framework providing lightweight bidirectional communication between your web clients and application services. Prysmo implemements a custom WebSocket protocol in one end and provides a fimiliar NodeJS API in the other.

## Get Started
- Install prysmo as a dependency in your gateway project: `npm i prysmo`
- In your code:

```js
// Import the module
const Prysmo = require('prysmo');

// Instance a new server
let server = new Prysmo();

// Register an endpoint
server.endpoint('hello', (sess, data, send) => send('Hello World!') );

// Start listening to default port
server.listen();
```

- Connect to the websocket at `ws://localhost:7667`
- Send the following message `{"endpoint":"hello"}`
- You will receive a message containing: `{"endpoint":"hello","data":"Hello World!"}`

## Features
- **WebSockets:** communicates with web clients through perforance oriented technology.
- **Encryption:** allows secure connections over TLS.
- **Session:** stores your users' credentials allowing seamless interaction with stateless services.
- **NodeJS:** gets out of your way delivering familiar development experience.
- **Logging:** emits a simple `log` event exposing all relevant information.
- **Settings:** all features farly configurable to fit in your use case.

## Client Implementations
- [Node Prysmo](https://gitlab.com/prysmo/node-prysmo)
- [JS Prysmo](https://gitlab.com/prysmo/js-prysmo)

## Archietcture

This is where prysmo will sit in your stack.

![Prymo Placement](https://i.imgur.com/mnjAYWM.png)

## Documentation
- [Quick Start]()
- [API Reference]()
- [Settings Reference]()
- [Usage Examples]()
- [Prism Protocol Spec]()

## Community
Meet us at:

- [Twitter](https://twitter.com/_prysmo)
- [Gitter](https://gitter.im/prysmo/community)


CHANGELOG.md
# Prysmo Changelog

All notable changes to Prysmo will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] (2019-01-04)

### Fixed
- Fixed ugly exception that used to pop up when you `.close` a closed instance. Now we throw a custom exception.

### Added
- We now send an error message to the client when, an error is thrown inside a [debug endpoint](TODO).
- Add the method [`prysmo.entity`](TODO) for registering multiple endpoints through a batch object.

## [0.0.1] (2018-12-31)

### Added
- First version of the framework.

[0.0.1]: https://gitlab.com/prysmo/prysmo/tags/v0.0.1
[0.0.2]: https://gitlab.com/prysmo/prysmo/tags/v0.0.2


test/spec.js
const wtf = require('wtfnode');
const fs = require('fs');
const chai = require('chai');
chai.use(require('chai-fs'));
const assert = chai.assert;

const delay = s => new Promise(done => setTimeout(done, s * 1000));

const Prysmo = require('../lib/main');
const WebSocket = require('ws');

describe('Prysmo', () => {

    after(() => wtf.dump());

    describe('::constructor', () => {

        it('should create the prysmo instance', () => {
            let p = new Prysmo();
            assert.equal(typeof p.log, 'object');
        });

        it('should read settings object', () => {
            let p = new Prysmo({ custom: true });
            assert(p.settings.custom);
        });

    });

    describe('#listen', () => {

        it('should listen on default port', done => {
            let p = new Prysmo();
            p.listen();

            let c = new WebSocket('ws://localhost:7667');
            c.on('open', () => {
                p.close();
                done();
            });
        });

    });

    describe('#close', () => {

        it('should stop all server components', done => {
            let p = new Prysmo();
            p.listen();
            p.close();

            let c = new WebSocket('ws://localhost:7667');
            c.on('error', e => {
                assert(e);
                done();
            });
        });

        it('should throw on closing a closed server', () => {
            let p = new Prysmo();
            p.listen();
            p.close();
            assert.throws( () => p.close() );
        });

    });

    describe('#endpoint', () => {

        let p;
        beforeEach(() => {
            p = new Prysmo();
            p.listen();
        });

        afterEach(() => p.close());

        it('should fail when receiving wrong arguments', () => {
            assert.throws(() => p.endpoint());
        });

        it('should register an endpoint', () => {
            p.endpoint('test1', (s, d, send) => send('TEST 1') );
            assert.equal(p.endpoints['test1'].listenerCount('trigger'), 1);
        });

        it('should register multiple listeners to endpoint', () => {
            p.endpoint('test1', () => true );
            p.endpoint('test1', () => true );
            p.endpoint('test1', () => true );
            assert.equal(p.endpoints['test1'].listenerCount('trigger'), 3);
        });

        it('should throw errors when debug flag is up', done => {
            p.endpoint('error', () => { throw new Error('test') }, true );
            p.endpoint('proxy', function(){  assert.throws(() => this.trigger('error')); done() });
            let c = new WebSocket('ws://localhost:7667', ['prysmo']);
            c.on('open', () => c.send('{"endpoint":"proxy"}') );
        });

        it('should send errors when debug flag is up', done => {
            p.endpoint('error', () => { throw new Error('test') }, true );
            p.endpoint('proxy', function(){  assert.throws(() => this.trigger('error')) });

            let c = new WebSocket('ws://localhost:7667', ['prysmo']);
            c.on('open', () => c.send('{"endpoint":"proxy"}') );
            c.on('message', data => {
                let d = JSON.parse(data);
                assert.equal(d.error, 'test');
                done();
            });
        });

    });

    describe('#entity', () => {

        let p;
        beforeEach(() => {
            p = new Prysmo();
            p.listen();
        });

        afterEach(() => p.close());

        it('should fail when receiving wrong arguments', () => {
            assert.throws( () => p.entity() );
        });

        it('should register endpoints', () => {
            p.entity({
                test1: (s, d, send) => send('TEST 1'),
                test2: (s, d, send) => send('TEST 2')
            });
            assert.equal(p.endpoints['test1'].listenerCount('trigger'), 1);
            assert.equal(p.endpoints['test2'].listenerCount('trigger'), 1);
        });

        it('should register endpoints with prefix', () => {
            p.entity({
                test1: (s, d, send) => send('TEST 1'),
                test2: (s, d, send) => send('TEST 2')
            }, 'Prefix.');
            assert.equal(p.endpoints['Prefix.test1'].listenerCount('trigger'), 1);
            assert.equal(p.endpoints['Prefix.test2'].listenerCount('trigger'), 1);
        });

    });

    describe('Settings', () => {

        it('should listen on custom port [port]', done => {
            let p = new Prysmo({ port: 50 });
            p.listen();

            let c = new WebSocket('ws://localhost:50');
            c.on('open', () => {
                p.close();
                done();
            });
        });

        it('should require HTTPS handshake [secureOnly]', done => {
            let p = new Prysmo({ secureOnly: true });
            p.listen();

            let c = new WebSocket('ws://localhost:7667');
            c.on('unexpected-response', (o, r) => {
                assert.equal(r.statusCode, 426);
                assert(r.headers.upgrade);
                p.close();
                done();
            });
        });

        it('should ping at custom rate to check connection health [timeout]', function(done){
            this.timeout(10e3);

            let p = new Prysmo({ timeout: 6e3 });
            p.listen();

            let c = new WebSocket('ws://localhost:7667');
            c.on('ping', () => c.pong(async () => {
                await delay(1);
                p.close();
                done();
            }) );
        });

        describe('Session Persistence [session]', () => {

            it('should save session before closing [session.persist]', done => {
                let p = new Prysmo({ session: { persist: true } });
                p.listen();
                p.session.object.right = true;
                p.close();
                assert.fileContentMatch('./session.json', /right/);
                fs.unlink('./session.json', () => done());
            });

            it('should save session when required [session.persist]', done => {
                let p = new Prysmo({ session: { persist: true } });
                p.listen();
                p.session.object.right = true;
                p.session.save();
                assert.fileContentMatch('./session.json', /right/);
                p.close();
                fs.unlink('./session.json', () => done());
            });

            it('should not do anything when asked to save without persist [session.persist]', () => {
                let p = new Prysmo();
                p.listen();
                p.session.object.right = true;
                p.session.save();
                assert.notPathExists('./session.json');
                p.close();
            });

            it('should load session upon opening [session.persist]', done => {
                let p = new Prysmo({ session: { persist: true } });
                p.listen();
                p.session.object.right = true;
                p.close();

                assert(!p.session);

                p.listen();
                assert(p.session.object.right);
                p.close();
                fs.unlink('./session.json', () => done());
            });

            it.skip('should save session regularly [session.backupInterval]', () => {

            });

            it('should save session on custom file [session.backupFile]', done => {
                let p = new Prysmo({ session: {
                    persist: true,
                    backupFile: './custom.json'
                } });
                p.listen();
                p.session.object.right = true;
                p.close();
                assert.fileContentMatch('./custom.json', /right/);
                fs.unlink('./custom.json', () => done());
            });

            it('should remove expired sessions [session.expires]', done => {
                let p = new Prysmo({ session: {
                    persist: true,
                    expires: 1
                } });
                p.listen();

                let c = new WebSocket('ws://localhost:7667');
                c.on('open', () => c.send('{"token":"null"}') );
                c.on('message', e => {
                    let t = JSON.parse(e).token;
                    p.session.object[t].expiryDate = new Date();
                    p.close();
                    p.listen();
                    assert(!p.session.object[t]);
                    p.close();
                    done();
                });
            });

        });

        describe('TLS & HTTPS [ssl]', () => {

            it('should connect securely through TLS', done => {
                let p = new Prysmo({ ssl: {
                    key: './test/res/localhost.key',
                    cert: './test/res/localhost.cert'
                } });
                p.listen();

                let c = new WebSocket(
                    'wss://localhost:7667',
                    { rejectUnauthorized: false }
                );
                c.on('open', () => {
                    p.close();
                    done();
                });
            });

        });

    });

    describe('Interaction', () => {

        let p, c;
        beforeEach(() => {
            p = new Prysmo();
            p.listen();
            c = new WebSocket('ws://localhost:7667', ['other', 'prysmo']);
        });

        afterEach(() => {
            p.close();
            c.terminate();
        });

        describe('Prysmo Protocol', () => {

            it('should ping to check connection health', function(done){
                this.timeout(5000);
                c.on('ping', () => c.pong(async () => {
                    await delay(1);
                    done();
                }) );
            });

            it('should handout a new token when asked', done => {
                c.on('open', () => c.send('{"token":"null"}') );
                c.on('message', e => {
                    let m = JSON.parse(e);
                    assert.equal(typeof m.token, 'string');
                    done();
                });
            });

            it('should create a new session when unknown token is sent', done => {
                c.on('open', () => c.send('{"token":"null"}') );
                c.on('message', e => {
                    let t = JSON.parse(e).token;
                    assert.equal(typeof p.session.object[t], 'object');
                    done();
                });
            });

            it('should recover the session when correct token is sent', done => {
                c.once('open', () => c.send('{"token":"null"}') );
                c.once('message', e => {
                    let t = JSON.parse(e).token;
                    c.terminate();

                    p.session.object[t].test = true;

                    c = new WebSocket('ws://localhost:7667');
                    c.once('open', () => c.send('{"token":"' + t + '"}') );

                    c.once('message', e => {
                        let t = JSON.parse(e).token;
                        assert(typeof p.session.object[t].test);
                        done();
                    });
                });
            });

            it('should pick prysmo protocol', done => {
                c.on('open', () => {
                    assert.equal(c.protocol, 'prysmo');
                    done();
                });
            });

        });

        it('should log warning when bad JSON was received', done => {
            let t = 0;
            p.on('log', l => {
                t++;
                if(t == 2){
                    assert.equal(l, 'DEBUG');
                    done();
                }
            });

            c.on('open', () => c.send('{endpoint":"error"}') );
        });

        it('should log errors in handlers', done => {
            p.endpoint('error', () => { throw new Error('gotcha!') });

            let t = 0;
            p.on('log', l => {
                t++;
                if(t == 3){
                    assert.equal(l, 'ERROR');
                    done();
                }
            });

            c.on('open', () => c.send('{"endpoint":"error"}') );
        });

        it('should log when no handlers apply to endpoint', done => {
            let t = 0;
            p.on('log', l => {
                t++;
                if(t == 3){
                    assert.equal(l, 'INFO');
                    done();
                }
            });

            c.on('open', () => c.send('{"endpoint":"unknown"}') );
        });

        describe('Endpoint Listeners', () => {

            it('should expose session, input and server data to listeners', done => {
                p.endpoint('hello', function(s, d){
                    assert.equal(typeof s, 'object');
                    assert.equal(d, 5);
                    assert.equal(typeof this.server.settings, 'object');
                    done();
                });

                c.on('open', () => c.send('{"endpoint":"hello","data":5}') );
            });

            it('should send message when asked by listener', done => {
                p.endpoint('hello', (s, d, send) => send('success'), true );

                c.on('open', () => c.send('{"endpoint":"hello","data":"5"}') );
                c.on('message', e => {
                    let m = JSON.parse(e);
                    assert.equal(m.data, 'success');
                    done();
                });
            });

            it('should trigger another endpoint when asked', done => {
                p.endpoint('hello', (s, d, send) => send('success') );
                p.endpoint('proxy', function(){  this.trigger('hello') });

                c.on('open', () => c.send('{"endpoint":"proxy"}') );
                c.on('message', e => {
                    let m = JSON.parse(e);
                    assert.equal(m.endpoint, 'hello');
                    assert.equal(m.data, 'success');
                    done();
                });
            });

        });

    });

});


test/dev/test.js
#!node

const fs = require('fs');
const toml = require('toml');
const Prysmo = require('../../lib/main');

(function(){
    try{
        let settings = toml.parse(fs.readFileSync('./test/dev/conf.toml'));
        let p = new Prysmo(settings);
        p.on('log', (level, msg) => console.log(new Date().toISOString().replace(/T/, ' ').replace(/\..+/, ''), level, msg) )

        p.endpoint(/hello/, (client, data, send) => send('HELLO!!'));

        function handle(){ p.close(); }

        process.on('SIGINT', handle);
        process.on('SIGTERM', handle);

        p.listen();
        console.log('%s listening at %s', p.name, p.port);
    }
    catch(e){
        console.log(e.message);
        console.log(e.stack || '');
    }
})();


lib/client.js

const Incoming = require('./incoming');
const Outgoing = require('./outgoing');

/*============================================================================o\
    Verify if connection with bound client's socket is still alive.
\o============================================================================*/
function checkHealth(){

    // If client has not responded with pong, close the socket.
    /* istanbul ignore if */
    if(this.isAlive === false)
        return this.kill();

    // Assume the connection is dead.
    this.isAlive = false;

    // Ping the client for activity.
    /* istanbul ignore else */
    if(this.socket.readyState == 1)
        this.socket.ping();
}

/*============================================================================o\
    This class represents a new WebSocket client to consume prysmo endpoints.
\o============================================================================*/
module.exports = class client{

    /*------------------------------------------------------------------------o\
        Constructor
    \o------------------------------------------------------------------------*/
    constructor(server, socket){
        this.log = server.log;
        this.ip = socket._socket.remoteAddress;
        this.isAlive = true;
        this.server = server;
        this.token = 'null';
        this.socket = socket;
        this.log.CONN_OPEN(this.ip);

        // Set the check alive routine.
        this.timeout = setInterval(checkHealth.bind(this), this.server.settings.timeout || 3e3);

        // Handle client websocket events.
        socket.on('pong', () => this.isAlive = true);
        socket.on('message', msg => new Incoming(this, msg));
        socket.on('error', /* istanbul ignore next */ err => this.server.emit('log', 'ERROR', err) );
        socket.on('close', () => this.kill());
    }

    /*------------------------------------------------------------------------o\
        [Endpoint API] Trigger the @endpoint's listeners inputing @data.
    \o------------------------------------------------------------------------*/
    trigger(endpoint, data){

        // Log and return if given endpoint doesn't exist.
        if(!this.server.endpoints[endpoint])
            return this.log.UNKNOWN_EP(endpoint);

        // Trigger endpoint.
        this.server.endpoints[endpoint].trigger(this, data);
    }

    /*------------------------------------------------------------------------o\
        Attempt recovering for client a session block associated with @token.
    \o------------------------------------------------------------------------*/
    claimSession(token){
        let ot = this.token;
        this.token = token;

        // Ask session to recover it.
        this.token = this.server.session.recover(this);
        this.server.emit('log', 'DEBUG', this.server.session.object);

        // Send the new token to client.
        Outgoing.token(this);
        this.log.TOKEN_RSET(ot, this.token);
    }

    /*------------------------------------------------------------------------o\
        Breaks the websocket connection between client and server.
    \o------------------------------------------------------------------------*/
    kill(){
        this.socket.terminate();
        this.log.CONN_CLOSE(this.token);
        clearInterval(this.timeout);
    }
}


lib/logger.js

/*============================================================================o\
    Centralize all text for logging.
\o============================================================================*/
let data = {
    MSG_RECVD: [ 'DEBUG', 'Received message ```%s´´´ from `%s´' ],
    MSG_SENT: [ 'DEBUG', 'Sent message ```%s´´´ to `%s´' ],
    BAD_JSON: [ 'DEBUG', 'Received malformed JSON ```%s´´´ from `%s´' ],
    CONN_DENIED: [ 'INFO', 'Connection denied for `%s´' ],
    TOKEN_RNEW: [ 'INFO', 'Connection `%s´ was named `%s´' ],
    TOKEN_RSET: [ 'INFO', 'Connection `%s´ was named `%s´' ],
    CONN_OPEN: [ 'INFO', 'Connection with `%s´ was established' ],
    CONN_CLOSE: [ 'INFO', 'Connection `%s´ was closed' ],
    UNKNOWN_EP: [ 'INFO', 'Could not handle enptoint `%s´' ]
    // Append new log messages as NAME: [ level, message ]
}

/*============================================================================o\
    Emit log event for parent, replacing placeholders in message.
\o============================================================================*/
function log(level, string, ...placeholders){
    let count = 0;
    let msg = string.replace(/%s/g, () => placeholders[count++]);
    this.emit('log', level, msg);
}

/*============================================================================o\
    Generate logging functions for all coded log messages.
\o============================================================================*/
module.exports = function(emitter){
    let o = {};
    for(let k in data)
        o[k] = log.bind(emitter, data[k][0], data[k][1]);
    return o;
};


lib/session.js

const fs = require('fs');

/*============================================================================o\
    Load bound session' JSON content from disk.
\o============================================================================*/
function loadSession(){
    try{
        return JSON.parse(fs.readFileSync(this.file, 'utf-8'));
    }
    catch(e){
        return {};
    }
}

/*============================================================================o\
    Execute expiry for @token of bound session.
\o============================================================================*/
function expireClient(token){

    // Remove session block if expiry date is past.
    if(new Date(this.object[token].expiryDate) <= new Date())
        this.remove(token);
}

/*============================================================================o\
    This class represents a session storage for persisting client's state
    inbetween connections.
\o============================================================================*/
module.exports = class Session{

    /*------------------------------------------------------------------------o\
        constructor
    \o------------------------------------------------------------------------*/
    constructor(settings){
        this.object = {};
        this.persist = Boolean(settings.persist);

        // Define persistence settings.
        if(this.persist){
            this.expires = settings.expires;
            this.file = settings.backupFile || 'session.json';
            this.object = loadSession.bind(this)();

            // Defines an interval to save current state to disk.
            this.interval = setInterval(() => this.save(), settings.backupInterval * 60e3);
        }
    }

    /*------------------------------------------------------------------------o\
        Save the state to disk and empty data store.
    \o------------------------------------------------------------------------*/
    close(){
        if(this.persist){
            clearInterval(this.interval);
            this.save();
        }
        this.object = {};
    }

    /*------------------------------------------------------------------------o\
        Save state to disk.
    \o------------------------------------------------------------------------*/
    save(){
        if(!this.persist)
            return;

        // Verivy expiry for every token.
        if(this.expires)
            Object.keys(this.object).forEach( t => expireClient.bind(this)(t) );

        // Write stored data to disk.
        fs.writeFileSync(this.file, JSON.stringify(this.object), 'utf-8');
    }

    /*------------------------------------------------------------------------o\
        Add a new session block for client.
    \o------------------------------------------------------------------------*/
    add(client){

        // Generate a new token.
        var nt = require('uuid/v4')();

        // Define the standard session object.
        this.object[nt] = {
            address: client.ip
        };

        // Renew block's expiry date.
        this.renew(nt);

        // Return new token.
        return nt;
    }

    /*------------------------------------------------------------------------o\
        Attempt recovering a previous session block due to @client's
        reconnection.
        > If a session block with client's token doesn't exist, a new empty
        > session block will be silently returned in order to difficult session
        > hijacking.
    \o------------------------------------------------------------------------*/
    recover(client){
        let token = client.token;

        // Add a new session block.
        let nt = this.add(client);

        // Check if client's previous block exists.
        if(token && this.object[token]){

            // Move block to new token space and delete old token.
            this.object[nt] = this.object[token];
            this.remove(token);
        }

        // Renew expiry date and update ip address.
        this.renew(nt);
        this.object[nt].address = client.ip;

        // Return new token.
        return nt;
    }

    /*------------------------------------------------------------------------o\
        Remove the block associated with @token.
    \o------------------------------------------------------------------------*/
    remove(token){
        this.object[token] = null;
        delete this.object[token];
    }

    /*------------------------------------------------------------------------o\
        Postpones @token's block expiry date according to settings.
    \o------------------------------------------------------------------------*/
    renew(token){
        var ed = new Date();
        ed.setHours(ed.getHours() + this.expires);
        this.object[token].expiryDate = ed;
    }
}


lib/endpoint.js
const EventEmitter = require('events');
const Outgoing = require('./outgoing');

/*============================================================================o\
    Instances of this class represent a API function or endpoint. They can be
    triggered to execute their associated logic. Endpoint listeners will usually
    interact with external services.
\o============================================================================*/
module.exports = class Endpoint extends EventEmitter{

    /*------------------------------------------------------------------------o\
        constructor
        - if @debug, listeners' errors will be thrown besides logged.
    \o------------------------------------------------------------------------*/
    constructor(name, debug){
        super();
        this.setMaxListeners(20);
        this.name = name;
        this.debug = debug || false;
    }

    /*------------------------------------------------------------------------o\
        Call all listeners for the 'trigger' event, passing @client and @data.
    \o------------------------------------------------------------------------*/
    trigger(client, data){
        this.emit('trigger', client, data);
    }

    /*------------------------------------------------------------------------o\
        Add @listener to the 'trigger' event, passing following context data:
        - this: client who triggered the endpoint
        - arg[0]: client's session object
        - arg[1]: input data sent by client
        - arg[2]: a function for the user to send responses to client
    \o------------------------------------------------------------------------*/
    register(listener){

        // Add the given listener to 'trigger' event.
        this.on('trigger', (client, data) => {

            // Grab session object for client.
            let sess = client.server.session.object[client.token] || {};

            // Bind send function with context data.
            let send = output => Outgoing.response(this, client, output);

            // Execute user listener passing context.
            try{
                listener.bind(client)(sess, data, send);
            }
            catch(e){

                // Log listeners' errors.
                client.server.emit('log', 'ERROR', e);
                if(this.debug){
                    Outgoing.error(client, e.message);
                    throw e;
                }
            }
        });

    }

}


lib/incoming.js

const Outgoing = require('./outgoing');

/*============================================================================o\
    When instnced, will parse @message and handle it for @client.
\o============================================================================*/
module.exports = class Incoming{

    /*------------------------------------------------------------------------o\
        constructor
    \o------------------------------------------------------------------------*/
    constructor(client, message){

        // Attempt parsing input message and log it.
        try{
            var msg = JSON.parse(message);
            client.log.MSG_RECVD(message, client.token);
        }

        // If message is invalid, send error to client.
        catch(e){
            Outgoing.error(client, 'I received a malformed JSON');
            return client.log.BAD_JSON(message, client.token);
        }

        // If it is a token request, claim the given session.
        if(typeof msg.token == 'string')
            client.claimSession(msg.token);

        // If it tries to reach an endpoint, trigger it.
        if(typeof msg.endpoint == 'string')
            client.trigger(msg.endpoint, msg.data);
    }

}


lib/main.js
const assert = require('assert');
const fs = require('fs');
const https = require('https');
const ws = require('ws');
const EventEmitter = require('events');
const getLogger = require('./logger');
const Session = require('./session');
const Client = require('./client');
const Endpoint = require('./endpoint');

/*============================================================================o\
    Check whether connection @info comply with security settings.
\o============================================================================*/
function handleSecurity(info, cb){
    let r = !this.settings.secureOnly || info.secure;
    cb(r, 426, 'Please, use HTTPS.', { 'Upgrade': 'TLS/1.1' });
    if(!r)
        this.log.CONN_DENIED(info.req.socket.remoteAddress);
}

/*============================================================================o\
    This class represents the Prysmo server and is the key component of the prysmo
    architecture.
\o============================================================================*/
module.exports = class Prysmo extends EventEmitter{

    /*------------------------------------------------------------------------o\
        Constructor
    \o------------------------------------------------------------------------*/
    constructor(settings){
        super();
        settings = settings || {};
        this.log = getLogger(this);
        this.settings = settings;
        this.name = 'Prysmo API Gateway';
        this.port = settings.port || 7667;
        this.endpoints = [];
    }

    /*------------------------------------------------------------------------o\
        Start listening for connections according to settings.
    \o------------------------------------------------------------------------*/
    listen(){

        // Start the session storage.
        this.session = new Session(this.settings.session || {});

        // Define initial websocket server settings.
        let wsOpts = {
            verifyClient: handleSecurity.bind(this),
            handleProtocols: () => 'prysmo'
        };

        // Start a HTTPS server if SSL was setup.
        if(this.settings.ssl && this.settings.ssl.key && this.settings.ssl.cert){
            wsOpts.server = https.createServer({
                key: fs.readFileSync(this.settings.ssl.key),
                cert: fs.readFileSync(this.settings.ssl.cert)
            });
            wsOpts.server.listen(this.settings.port || 7667);
        }
        else
            wsOpts.port = this.settings.port || 7667;

        // Start WebSocket server according to settings.
        this.server = new ws.Server(wsOpts);
        this.server.on('connection', socket => new Client(this, socket));
        this.server.on('error', /* istanbul ignore next */ e => this.emit('log', 'ERROR', e) );
    }

    /*------------------------------------------------------------------------o\
        Stop all server operations and reset it's state.
    \o------------------------------------------------------------------------*/
    close(){

        // Throw if Prysmo was already closed.
        if(!this.server)
            throw new Error('Cannot close a closed Prysmo');

        // Stop all server components.
        this.server._server.close();
        this.server.close();
        this.session.close();
        this.session = null;
        this.server = null;
    }

    /*------------------------------------------------------------------------o\
        Add @listener to @name endpoint.
    \o------------------------------------------------------------------------*/
    endpoint(name, listener, debug){
        assert.equal(typeof name, 'string');
        assert.equal(typeof listener, 'function');

        // Create endpoint if it did not exist.
        if(!this.endpoints[name])
            this.endpoints[name] = new Endpoint(name, debug);

        // Add listener to given endpoint.
        this.endpoints[name].register(listener);
    }

    /*------------------------------------------------------------------------o\
        Register key-value pairs of @spec as endpoint-listener. If @prefix,
        prepend it to all endpoint names.
    \o------------------------------------------------------------------------*/
    entity(spec, prefix){
        assert.equal(typeof spec, 'object');
        let p = prefix ? String(prefix) : '';
        for(let k in spec)
            this.endpoint(p + k, spec[k]);
    }

}
